### PR TITLE
#339 introduce resource bundle

### PR DESCRIPTION
**Pull request by sjaakd**
I took the approach of introducing a resource bundle iso the constants as proposed in #339. Also fixed some small inconsistencies (interpunction, one tiny spelling error). 

It shows nicely how messages / terminology are used. We could be more consistent with quotes (sometimes we use them, sometimes not). But all-in-all, its not that bad :smile: 

`String.format` a re-occuring pattern, the solution pushes the `String.format` to the newly introduced `FormattingMessager`.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Wow, that seems like a an awful lot of work :)

**Comment by sjaakd** [comment]
Ha! Now the stats work I want to be on top :smile:. Without kidding, it's not so bad. Just a bit boring and tedious. Costed me 3 hours in the end.

**Comment by gunnarmorling** [comment]
@sjaakd, I think the centralization of messages is a great idea. But maybe we can do better than using Strings for referencing messages.

E.g. we can use an enum as in the top answer in this [SO post](https://stackoverflow.com/questions/446663/best-way-to-define-error-codes-strings-in-java). We can have the messages within the enum type itself, as we don't want to I18N messages, so there is no strong benefit to using a resource bundle. Such approach will avoid typos when referencing messages and also makes it a bit more refactoring-safe.

WDYT?

**Comment by gunnarmorling** [comment]
> Ha! Now the stats work I want to be on top

Ah, nice that it's working now. Indeed I need to keep an eye on you ;)

**Comment by gunnarmorling** [comment]
Or maybe an approach like this: https://code.google.com/p/jlibs/wiki/Internationalization

I wouldn't want to add a dependency to an external library, but we could easily implement such an approach ourselves e.g. using a dynamic proxy (or via code generation using an annotation processor :-) ).

**Comment by sjaakd** [comment]
> E.g. we can use an enum as in the top answer in this SO post. We can have the messages within the enum type itself, as we don't want to I18N messages, so there is no strong benefit to using a resource bundle. Such approach will avoid typos when referencing messages and also makes it a bit more refactoring-safe.
> 
> WDYT?

I think that's a good idea.. However, I would like to switch off the 120 characters limit then on the definition part of this particular enum. I think it offers quite nice insight to have 'unbroken' strings (at least it did for me)..

**Comment by gunnarmorling** [comment]
> However, I would like to switch off the 120 characters limit

Yes, I guess that makes sense.

**Comment by sjaakd** [comment]
Lets go for the enum approach then.. Keep it simple, plain and stupid :+1:

**Comment by gunnarmorling** [comment]
Ok :)

**Comment by sjaakd** [comment]
Ok.. Done!

File: processor/src/main/java/org/mapstruct/ap/model/source/IterableMapping.java [hunk 1]
```diff
@@ -52,12 +53,7 @@ public static IterableMapping fromPrism(IterableMappingPrism iterableMapping, Ex
         if ( !elementTargetTypeIsDefined
             && iterableMapping.dateFormat().isEmpty()
             && iterableMapping.qualifiedBy().isEmpty() ) {
-            messager.printMessage(
-                Diagnostic.Kind.ERROR,
-                "'dateformat', 'qualifiedBy' and 'elementTargetType' are undefined in @IterableMapping, "
-                    + "define at least one of them.",
-                method
-            );
+            messager.printMessage( Diagnostic.Kind.ERROR, method, Message.iterablemapping_noelements );
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/model/source/IterableMapping.java]: Now with the enum approach in place, couldn't also the severity/kind be part of the enum definition? I don't think one message is reported with different kinds at different places, so we can move it there and simplify the calls to printMessage().

File: processor/src/main/java/org/mapstruct/ap/util/Message.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+package org.mapstruct.ap.util;
+
+/**
+ *
+ * @author Sjaak Derksen
+ */
+public enum Message {
+
+    // CHECKSTYLE:OFF
+    beanmapping_noelements( "'resultType' and 'qualifiedBy' are undefined in @BeanMapping, define at least one of them." ),
```
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/util/Message.java]: Any chance you convert the enum member names to upper-case? It's just more common and what you'd expect from it. Not sure though whether there is a good way to automate?

File: processor/src/test/java/org/mapstruct/ap/model/common/DefaultConversionContextTest.java [hunk 3]
```diff
@@ -119,26 +120,26 @@ private Type typeWithFQN(String fullQualifiedName) {
...
     }
 
-    private static class StatefulMessagerMock implements Messager {
+    private static class StatefulMessagerMock implements FormattingMessager {
 
         private Diagnostic.Kind lastKindPrinted;
 
         @Override
-        public void printMessage(Diagnostic.Kind kind, CharSequence msg) {
+        public void printMessage( Diagnostic.Kind kind, Message msg, Object... arg) {
```
Comment (gunnarmorling) [inline processor/src/test/java/org/mapstruct/ap/model/common/DefaultConversionContextTest.java]: We don't do a whitespace after braces in method definitions usually.

Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/model/common/DefaultConversionContextTest.java]: Ok..

File: processor/src/main/java/org/mapstruct/ap/util/Message.java (hunk 2 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/util/Message.java]: Yep.. I'll just do the refactor mode.. A bit of effort..

File: processor/src/main/java/org/mapstruct/ap/model/source/IterableMapping.java (hunk 1 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/model/source/IterableMapping.java]: Good point..

Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/model/source/IterableMapping.java]: Maybe we can add a CS check for it?

**Comment by sjaakd** [comment]
@gunnarmorling , processed your comments into an additional commit. Please have a look.

**Comment by gunnarmorling** [comment]
Merged; Many thanks, @sjaakd! It's much nicer with the enum now :+1:

**Pull request merged**